### ISSUE TITLE
No implementation found for java.lang.String dalvik.system.DexFile.getOatFileCompilerFilter() in OOS

### ISSUE THREAD

**Issue by aviraxp** [rovo89/XposedBridge#203]
This has been reported by many, but no one wants to provides a log. Hopefully someone has a module for it to fix (http://repo.xposed.info/module/com.fifsource.android.oosbackgroupdexcompilationfix), and the code decompiled looks like this:

```
public void handleLoadPackage(LoadPackageParam lpParam) {
        if (lpParam.packageName.equals("android")) {
            try {
                Class pdoClass = XposedHelpers.findClass("com.android.server.pm.PackageDexOptimizer", lpParam.classLoader);
                final Field packagePackageNameField = XposedHelpers.findField(XposedHelpers.findClass("android.content.pm.PackageParser$Package", lpParam.classLoader), "packageName");
                Class packageManagerServiceClass = XposedHelpers.findClass("com.android.server.pm.PackageManagerService", lpParam.classLoader);
                XposedHelpers.setStaticBooleanField(packageManagerServiceClass, "DEBUG_DEXOPT", true);
                XposedBridge.hookMethod(XposedHelpers.findMethodExact(pdoClass, "getOatFileCompilerFilter", new Class[]{packageClass}), new XC_MethodHook() {
                    protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                        XposedBridge.log("OOSDEX: getOatFileCompilerFilter(\"" + ((String) packagePackageNameField.get(param.args[0])) + "\") = \"" + null + "\"");
                        param.setResult(null);
                    }
                });
                XposedBridge.hookMethod(XposedHelpers.findMethodExact(packageManagerServiceClass, "performDexOpt", new Class[]{String.class, Boolean.TYPE, Integer.TYPE, Boolean.TYPE}), new XC_MethodHook() {
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        String packageName = param.args[0];
                        boolean checkProfile = ((Boolean) param.args[1]).booleanValue();
                        int compileReason = ((Integer) param.args[2]).intValue();
                        boolean force = ((Boolean) param.args[3]).booleanValue();
                        XposedBridge.log("OOSDEX: performDexOpt(\"" + packageName + "\", " + checkProfile + ", " + compileReason + ", " + force + ") = " + ((Boolean) param.getResult()).booleanValue());
                    }
                });
                XposedBridge.log("OOSDEX: Hooked successfully");
            } catch (ClassNotFoundError e) {
                XposedBridge.log("OOSDEX: Failed to hook, is that an OOS ROM?");
            }
        }
    }

```
Could you take a look at that and maybe include it in XposedBridge?

**Comment by rovo89** [comment]
@wanam mentioned the name getOatFileCompilerFilter() to me before, as a difference between SDK24 and SDK25. He also mentioned DexFile.getDexFileOutputPath(), which has been added in SDK25 with this commit: https://android.googlesource.com/platform/art/+/b4827ace453b9280060a826e8f22cc8c9b6edb7d%5E!/

Questions:

1. Is this crash on Android 7.0 or 7.1 or both?
2. Could you please provide the smali code for com.android.server.pm.PackageDexOptimizer and dalvik.system.DexFile? Both seem to be modified in OOS, the getOatFileCompilerFilter() doesn't exist in either class in AOSP.
3. Could you please provide their /system/lib/libart.so?

**Comment by aviraxp** [comment]
https://github.com/joshuous/op5_system_dump

Someone has uploaded here. It is 7.1.1, but no one can confirm if 7.0 has this issue because no one uses that now. I can turn it into smali if you need.

But we do not know which version this issue comes from, because OP3 users are complaining too and there are too many versions of oos.

**Comment by rovo89** [comment]
> I can turn it into smali if you need.

That would help, yes. I could obviously do that myself, but as I'm very busy, the saved timed can be spent on actually fixing things. :)

**Comment by aviraxp** [comment]
https://github.com/aviraxp/for_rovo89

All there.

**Comment by rovo89** [comment]
Thanks! I had started to implement the native method and was wondering what should happen in error situations (return null? empty string? throw exception?) From the declaration, I know now that I need to look out for a FileNotFoundException.

**Comment by rovo89** [comment]
Could you please try this? https://www.dropbox.com/s/abmb1eywfleau6z/xposed-v88.2-sdk25-arm64-test2.zip?dl=0
In the native code, I didn't find the FileNotFoundException, but I could be wrong, and considering that it would make sense, the method declaration includes it and PackageDexOptimizer catches for IOExceptions, it should be the right way to do it.

**Comment by aviraxp** [comment]
Unfortunately it does not work. Cannot provide logs because the person who test it charged at night and logs are replaced by newer ones. How about let it return null instead, like that module?

**Comment by rovo89** [comment]
I prefer fixes over workarounds. Returning null can have unknown side effects, the caller probably wants to know the compiler filter for a good reason and might fall back to less optimal behavior when it can't get it.

So I'll wait for new logs, please let me know once you have them.

**Comment by rovo89** [comment]
Did you hear anything further from that person? @wanam verified that at least calling the method doesn't crash and that it returns sensible values (at least for certain apps).

**Comment by aviraxp** [comment]
Nah, seems nobody wants to provide a log as there is an available module to fix.

**Comment by rovo89** [comment]
Great... why did I even spend some hours on this then? Not your fault, I know, just saying...

By the way, the module is a workaround, not a fix. It will always pretend that the oat file doesn't exist (by returning `null`). As far as I could see, the `PackageDexOptimizer` triggers a full recompile in this case. I haven't looked further into it, but possibly this might happen every day. So what I'll probably do is include the implementation as it is now, I think it's close to the original one.

**Comment by rovo89** [comment]
(Hopefully) fixed in v88.2.
